LeaderboardBackend/Services/Impl/EmailSender.cs

```diff
@@ -0,0 +1,62 @@
+using MailKit.Net.Smtp;
+using Microsoft.Extensions.Options;
+using MimeKit;
+
+namespace LeaderboardBackend.Services;
+
+public class EmailSender : IEmailSender
+{
+    private readonly EmailSenderConfig _config;
+    private readonly ILogger<EmailSender> _logger;
+    private readonly ISmtpClient _smtpClient;
+    private readonly MailboxAddress _sender;
+
+    public EmailSender(IOptions<EmailSenderConfig> config, ILogger<EmailSender> logger, ISmtpClient smtpClient)
+    {
+        _config = config.Value;
+        _logger = logger;
+        _smtpClient = smtpClient;
+        _sender = new MailboxAddress(name: _config.SenderName, address: _config.SenderAddress);
+    }
+
+    public async Task SendEmailAsync(string recipientAddress, string subject, string htmlMessage)
+    {
+        if (_config.Smtp is null)
+        {
+            _logger.LogWarning("Can't send email (subject: \"{EmailSubject}\"), SMTP configuration is missing",
+                subject);
+            return;
+        }
+
+        MimeMessage message = new();
+        message.From.Add(_sender);
+        message.To.Add(new MailboxAddress(name: null, address: recipientAddress));
+        message.Subject = subject;
+
+        BodyBuilder bodyBuilder = new() { HtmlBody = htmlMessage };
+        message.Body = bodyBuilder.ToMessageBody();
+
+        try
+        {
+            await SendEmail(message);
+            _logger.LogInformation("Sent email (subject \"{EmailSubject}\")", subject);
+        }
+        catch (Exception exception)
+        {
+            _logger.LogError(exception, "Caught exception when sending email (subject: \"{EmailSubject}\")", subject);
+        }
+    }
+
+    private async Task SendEmail(MimeMessage message)
+    {
+        await _smtpClient.ConnectAsync(_config.Smtp!.Host, _config.Smtp.Port, _config.Smtp.UseSsl);
```

Comment:
Is it a good idea to keep this connection open or just connect ad hoc like this? I am a bit ignorant on best practices here.

Reply:
a single SMTP instance is now reused as talked about on Discord